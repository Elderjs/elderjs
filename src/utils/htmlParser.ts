import {parseExpressionAt} from "acorn";

type Node = {
  start: number;
  end: number;
}

type TagNode = Node & {
  name: string;
  attrs: Array<AttrNode|ExpressionNode>;
  selfClosed: boolean;
  start: number;
  end: number;
}

type AttrNode = Node & {
  name: string;
  value?: AttrValueNode;
}

type AttrValueNode = Node & {
  value?: string;
  raw?: string;
}

type ExpressionNode = Node & {
  exp: any;
  spread: boolean;
}

export function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#039;');
}

export const unescapeHtml = (str) =>
  str
    .replace(/\\\\n/gim, '\\n')
    .replace(/&quot;/gim, '"')
    .replace(/&lt;/gim, '<')
    .replace(/&gt;/gim, '>')
    .replace(/&#0?39;/gim, "'")
    .replace(/\\"/gim, '"')
    .replace(/&amp;/gim, '&');

export function parseExpression(content: string, index: number): ExpressionNode {
  const rx = /{\s*(\.\.\.)?/y;
  rx.lastIndex = index;
  const [prefix,, hasSpread] = rx.exec(content);
  const exp = parseExpressionAt(content, index + prefix.length);
  const rxEnd = /\s*}/y;
  rxEnd.lastIndex = exp.end;
  rxEnd.exec(content);
  return {
    start: index,
    end: rxEnd.lastIndex,
    exp,
    spread: Boolean(hasSpread)
  };
}

export function parseAttrValue(content: string, index: number): AttrValueNode | ExpressionNode {
  let raw;
  let rx;
  if (content[index] === "'") {
    rx = /'([^']*)'/y;
    rx.lastIndex = index;
    raw = content.match(rx)[1];
  } else if (content[index] === '"') {
    rx = /"([^"]*)"/y;
    rx.lastIndex = index;
    raw = content.match(rx)[1];
  } else if (content[index] === "{") {
    return parseExpression(content, index);
  } else {
    rx = /[^\s"'=<>\/\x7f-\x9f]+/y;
    rx.lastIndex = index;
    raw = content.match(rx)[0];
  }
  return {
    value: unescapeHtml(raw),
    raw,
    start: index,
    end: rx.lastIndex
  };
}

export function parseTag(content: string, index: number): TagNode {
  const rx = /<([\w-:]+)/y;
  rx.lastIndex = index;
  const match = content.match(rx);
  const attrs = parseAttrs(content, rx.lastIndex);
  const rxEnd = /\s*(\/?)>/y;
  if (attrs.length) {
    rxEnd.lastIndex = attrs[attrs.length - 1].end;
  } else {
    rxEnd.lastIndex = rx.lastIndex;
  }
  const matchEnd = content.match(rxEnd);
  return {
    name: match[1],
    attrs,
    selfClosed: Boolean(matchEnd[1]),
    start: index,
    end: rxEnd.lastIndex
  };
}

export function parseAttrs(content: string, index: number): Array<AttrNode|ExpressionNode> {
  const rx = /(\s+)(?:([\w-:]+)(\s*=\s*)?|{)/y;
  rx.lastIndex = index;
  const result = [];
  let match;
  while ((match = rx.exec(content))) {
    const [, prefix, name, hasValue] = match;
    if (!name) {
      // expression
      const node = parseExpression(content, match.index + prefix.length);
      result.push(node);
      rx.lastIndex = node.end;
      continue;
    }
    const value = hasValue ? parseAttrValue(content, rx.lastIndex) : null;
    result.push({
      start: match.index,
      end: value ? value.end : rx.lastIndex,
      name,
      value
    });
    if (value) {
      rx.lastIndex = value.end;
    }
  }
  return result;
}

