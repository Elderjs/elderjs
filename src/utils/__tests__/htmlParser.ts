/* eslint-disable no-param-reassign */
import { escapeHtml, unescapeHtml, parseTag } from '../htmlParser';

const cases = [['&amp;&#039;&quot;', `&'"`]];

describe('escapeHtml/unescapeHtml', () => {
  for (const [left, right] of cases) {
    test(`${left} ${right}`, () => {
      expect(escapeHtml(right)).toEqual(left);
      expect(unescapeHtml(left)).toEqual(right);
    });
  }
});

test('#escapeHtml', () => {
  expect(escapeHtml('')).toEqual('');
  expect(escapeHtml(`<html>'Tom'&amp;"Jerry"</html>`)).toEqual(
    '&lt;html&gt;&#039;Tom&#039;&amp;amp;&quot;Jerry&quot;&lt;/html&gt;',
  );
});

test('#unescapeHtml', () => {
  expect(unescapeHtml('{&quot;nh_count&quot;:15966,&quot;classes&quot;:&quot;mt-3&quot;}')).toEqual(
    '{"nh_count":15966,"classes":"mt-3"}',
  );
  expect(unescapeHtml('&quot;&lt;&gt;&#39;&quot;\\n\\\\n\\"&amp;')).toEqual('"<>\'"\\n\\n"&');
  expect(unescapeHtml('abcd 1234 <&""&>')).toEqual('abcd 1234 <&""&>');
});

describe('#parseTag', () => {
  test('basic tag', () => {
    expect(parseTag(`<div class="foo">`, 0)).toMatchSnapshot();
  });
  test('basic self closed tag', () => {
    expect(parseTag(`<div class="foo"/>`, 0)).toMatchSnapshot();
  });
  test('svelte value', () => {
    expect(parseTag(`<MyComponent class="foo" hydrate-client={props}>`, 0)).toMatchSnapshot();
  });
});
