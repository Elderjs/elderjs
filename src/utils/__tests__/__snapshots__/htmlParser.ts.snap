// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`#parseTag basic self closed tag 1`] = `
Object {
  "attrs": Array [
    Object {
      "end": 16,
      "name": "class",
      "start": 5,
      "value": Object {
        "end": 16,
        "raw": "foo",
        "start": 11,
        "value": "foo",
      },
    },
  ],
  "end": 18,
  "name": "div",
  "selfClosed": true,
  "start": 0,
}
`;

exports[`#parseTag basic tag 1`] = `
Object {
  "attrs": Array [
    Object {
      "end": 16,
      "name": "class",
      "start": 5,
      "value": Object {
        "end": 16,
        "raw": "foo",
        "start": 11,
        "value": "foo",
      },
    },
  ],
  "end": 17,
  "name": "div",
  "selfClosed": false,
  "start": 0,
}
`;

exports[`#parseTag svelte value 1`] = `
Object {
  "attrs": Array [
    Object {
      "end": 24,
      "name": "class",
      "start": 13,
      "value": Object {
        "end": 24,
        "raw": "foo",
        "start": 19,
        "value": "foo",
      },
    },
    Object {
      "end": 47,
      "name": "hydrate-client",
      "start": 25,
      "value": Object {
        "end": 47,
        "exp": Node {
          "end": 46,
          "name": "props",
          "start": 41,
          "type": "Identifier",
        },
        "spread": false,
        "start": 40,
      },
    },
  ],
  "end": 48,
  "name": "MyComponent",
  "selfClosed": false,
  "start": 0,
}
`;
